### ISSUE TITLE
Use reduced-size sample set for bootstrap estimation (delta, sobol)

### ISSUE THREAD

**Issue by simonmarwitz** [SALib/SALib#570]
I use a fairly large pre-generated sample set and use the Delta Moment-Independent Analysis for global sensitivity indices. I want to use only a subset of samples for estimation of delta and S1 indices and the bootstrap estimation of their confidence intervals. Therefore, I have introduced a new Parameter 'Nr' into the routines. Another minor improvement is an switch to deselect additional estimation of first-order sobol indices. As I have trouble creating a fork and a pull request, I attach my code in the hope, that on of you maintainer might merge it into the code base.

```python
from typing import Dict
from scipy.stats import norm, gaussian_kde, rankdata

import numpy as np

from . import common_args
from ..util import read_param_file, ResultDict


def analyze(
    problem: Dict,
    X: np.ndarray,
    Y: np.ndarray,
    num_resamples: int = 100,
    conf_level: float = 0.95,
    print_to_console: bool = False,
    seed: int = None,
    Nr: int = None,
    analyze_sobol: bool = True,
) -> Dict:
    """Perform Delta Moment-Independent Analysis on model outputs.

    Returns a dictionary with keys 'delta', 'delta_conf', 'S1', and 'S1_conf' 
    (first-order sobol indices), where each entry is a list of size D 
    (the number of parameters) containing the indices in the same order as the 
    parameter file. 


    Notes
    -----
    Compatible with:
        all samplers


    Examples
    --------
        >>> X = latin.sample(problem, 1000)
        >>> Y = Ishigami.evaluate(X)
        >>> Si = delta.analyze(problem, X, Y, print_to_console=True)


    Parameters
    ----------
    problem : dict
        The problem definition
    X: numpy.matrix
        A NumPy matrix containing the model inputs
    Y : numpy.array
        A NumPy array containing the model outputs
    num_resamples : int
        The number of resamples when computing confidence intervals (default 100)
    conf_level : float
        The confidence interval level (default 0.95)
    print_to_console : bool
        Print results directly to console (default False)
    Nr: int
        Number of samples to use when resampling (bootstrap) (default None)
    analyze_sobol: int
        Also compute first-order sobol indices and their confidence intervals (default True)


    References
    ----------
    1. Borgonovo, E. (2007). "A new uncertainty importance measure."
           Reliability Engineering & System Safety, 92(6):771-784,
           doi:10.1016/j.ress.2006.04.015.

    2. Plischke, E., E. Borgonovo, and C. L. Smith (2013). "Global
           sensitivity measures from given data." European Journal of
           Operational Research, 226(3):536-550, doi:10.1016/j.ejor.2012.11.047.
    """
    if seed:
        np.random.seed(seed)

    D = problem["num_vars"]
    if Nr is None:
        Nr = Y.size

    if not 0 < conf_level < 1:
        raise RuntimeError("Confidence level must be between 0-1.")

    # equal frequency partition
    exp = 2.0 / (7.0 + np.tanh((1500.0 - Nr) / 500.0))
    M = int(np.round(min(int(np.ceil(Nr**exp)), 48)))
    m = np.linspace(0, Nr, M + 1)
    Ygrid = np.linspace(np.min(Y), np.max(Y), 100)

    keys = ("delta", "delta_conf", "S1", "S1_conf")
    S = ResultDict((k, np.zeros(D)) for k in keys)
    S["names"] = problem["names"]

    try:
        for i in range(D):
            X_i = X[:, i]
            S["delta"][i], S["delta_conf"][i] = bias_reduced_delta(
                Y, Ygrid, X_i, m, num_resamples, conf_level, Nr
            )
            if analyze_sobol:
                ind = np.random.randint(Y.size, size=Nr)
                S["S1"][i] = sobol_first(Y[ind], X_i[ind], m)
                S["S1_conf"][i] = sobol_first_conf(Y, X_i, m, num_resamples, conf_level, Nr)
    except np.linalg.LinAlgError as e:
        msg = "Singular matrix detected\n"
        msg += "This may be due to the sample size ({}) being too small\n".format(
            Y.size
        )
        msg += "If this is not the case, check Y values or raise an issue with the\n"
        msg += "SALib team"

        raise np.linalg.LinAlgError(msg) from e

    if print_to_console:
        print(S.to_df())

    return S


def calc_delta(Y, Ygrid, X, m):
    """Plischke et al. (2013) delta index estimator (eqn 26) for d_hat."""
    N = len(Y)
    fy = gaussian_kde(Y, bw_method="silverman")(Ygrid)
    xr = rankdata(X, method="ordinal")

    d_hat = 0.0
    l_m = len(m) - 1
    for j in range(l_m):
        ix = np.where((xr > m[j]) & (xr <= m[j + 1]))[0]
        nm = len(ix)

        # if not np.all(np.equal(Y_ix, Y_ix[0])):
        Y_ix = Y[ix]
        if Y_ix.ptp() != 0.0:
            fyc = gaussian_kde(Y_ix, bw_method="silverman")(Ygrid)
            fy_ = np.abs(fy - fyc)
        else:
            fy_ = np.abs(fy)

        d_hat += (nm / (2 * N)) * np.trapz(fy_, Ygrid)

    return d_hat


def bias_reduced_delta(Y, Ygrid, X, m, num_resamples, conf_level, Nr):
    """Plischke et al. 2013 bias reduction technique (eqn 30)"""
    d = np.empty(num_resamples)

    N = len(Y)
    if Nr is None: 
        Nr = N
    ind = np.random.randint(N, size=Nr)
    d_hat = calc_delta(Y[ind], Ygrid, X[ind], m)
    r = np.random.randint(N, size=(num_resamples, Nr))
    for i in range(num_resamples):
        r_i = r[i, :]
        d[i] = calc_delta(Y[r_i], Ygrid, X[r_i], m)

    d = 2.0 * d_hat - d
    return (d.mean(), norm.ppf(0.5 + conf_level / 2) * d.std(ddof=1))


def sobol_first(Y, X, m):
    # pre-process to catch constant array
    # see: https://github.com/numpy/numpy/issues/9631
    if Y.ptp() == 0.0:
        # Catch constant results
        # If Y does not change then it is not sensitive to anything...
        return 0.0

    xr = rankdata(X, method="ordinal")
    Vi = 0
    N = len(Y)
    Y_mean = Y.mean()
    for j in range(len(m) - 1):
        ix = np.where((xr > m[j]) & (xr <= m[j + 1]))[0]
        nm = len(ix)
        Vi += (nm / N) * ((Y[ix].mean() - Y_mean) ** 2)

    return Vi / np.var(Y)


def sobol_first_conf(Y, X, m, num_resamples, conf_level, Nr):
    s = np.zeros(num_resamples)

    N = len(Y)
    if Nr is None: 
        Nr = N
    r = np.random.randint(N, size=(num_resamples, Nr))
    for i in range(num_resamples):
        r_i = r[i, :]
        s[i] = sobol_first(Y[r_i], X[r_i], m)

    return norm.ppf(0.5 + conf_level / 2) * s.std(ddof=1)


def cli_parse(parser):
    parser.add_argument(
        "-X",
        "--model-input-file",
        type=str,
        required=True,
        default=None,
        help="Model input file",
    )
    parser.add_argument(
        "-r",
        "--resamples",
        type=int,
        required=False,
        default=10,
        help="Number of bootstrap resamples for \
                           Sobol confidence intervals",
    )
    return parser


def cli_action(args):
    problem = read_param_file(args.paramfile)
    Y = np.loadtxt(
        args.model_output_file, delimiter=args.delimiter, usecols=(args.column,)
    )
    X = np.loadtxt(args.model_input_file, delimiter=args.delimiter, ndmin=2)
    if len(X.shape) == 1:
        X = X.reshape((len(X), 1))

    analyze(
        problem,
        X,
        Y,
        num_resamples=args.resamples,
        print_to_console=True,
        seed=args.seed,
    )


if __name__ == "__main__":
    common_args.run_cli(cli_parse, cli_action)

```

**Comment by tupui** [comment]
Hi @simonmarwitz, thank you for the detailed report.

Both requests are interesting, but I am not sure about the API.

For `analyze_sobol`, I would actually prefer that we had a `method` parameter. There are many metrics that we could plug there, not just delta and Sobol'.

For the number of resampling used for confidence intervals, I would also move all of that to a result object. For a bunch of reasons including recomputing CIs, on SciPy we are moving towards that direction. I believe we could do the same here.

Furthermore, in SciPy we recently introduced `bootstrap` which provides a very powerful interface with a few methods including bias reduction as present here.

TL;DR, I would propose instead of updating the current API that we rework the API from the ground up. We had some short discussions about sharing some codes from a few methods working in a similar fashion as moment independent (bins.)

**Comment by simonmarwitz** [comment]
Hi @tupui, thanks for your quick response.

As I am just applying SALib for my research, unfortunately I don't have the resources to implement the proposed major changes. I simply shared my modifications here, so others could benefit from it. Nevertheless I can totally see your points. 

The `analyze_sobol` switch was added by me, to be able to actually **disable** computation of sobol indices, as it was enabled before (hardcoded). I was confused in the beginning, why they were included at all in the result object, as there was no explanation in the docs about it. 

> For the number of resampling used for confidence intervals, I would also move all of that to a result object. 

That, I don't understand. Why would you put a function argument into a result object? Anyway, it certainly makes sense to use and improve SciPy's `bootstrap` instead of self-implementing it everywhere in SALib. 

PS: I did not know, you were the people behind SciPy `bootstrap`. I am currently using it for other things and i am working on some improvements; might open an issue in SciPy, when I feel they are ready.

**Comment by ConnectedSystems** [comment]
Thanks @simonmarwitz 

@tupui I agree with your comments in general but I feel they can be addressed later, perhaps as part of a larger look and evaluation of the SALib architecture.

As a proponent of incremental improvement, I think it would be worth putting this in as is.

@simonmarwitz I would be happy to submit a PR for this on your behalf but if would be a shame for you not to get recognized as the contributor. Could you elaborate on what issue you had forking and submitting a PR? Perhaps it's something we can help with?

**Comment by tupui** [comment]
Sure we can iterate. If folks are ok to update their codes twice, I have no objections (I even prefer that tbh, just coming from a SciPy perspective where backward compatibility and churn concerns is pushed to the extreme) 😃

> That, I don't understand. Why would you put a function argument into a result object? 

I somehow agree that it does not necessarily make sense to have methods embedded in a result object. The main line, idea, is to have some functionalities outside of the main function so you don't need to recompute everything to just update a param like a CI threshold for instance.

> PS: I did not know, you were the people behind SciPy bootstrap. I am currently using it for other things and i am working on some improvements; might open an issue in SciPy, when I feel they are ready.

Credit goes to Matt, but I did participate 😉 I am looking forward to reading this! FYI @mdhaber

I am also happy to help for any on-boarding related issues.

**Comment by ConnectedSystems** [comment]
Closed by #571 

Thanks again @simonmarwitz
